README.md
# Sticky Postbox #
**Contributors:** Enrico Sorcinelli  
**Tags:** Administration, Dashboard, Options, Sticky, Post, Classic Editor, meta box  
**Requires at least:** 4.4  
**Requires PHP:** 5.2.4  
**Tested up to:** 6.0  
**Stable tag:** 1.3.0  
**License:** GPLv2 or later  

Add sticky feature to administration meta boxes.

## Description ##

**Sticky Postbox** is a piece of code written few years ago as exercise to learn how WordPress handles administration meta boxes's availables statuses (closed, hidden and sortings).

This very lightweight plugin adds the _sticky_ feature to administration meta boxes allowing them to be sticky at top right corner of the browser window.

Only one meta box can be sticky at once, so a new sticky meta box unstick the current one (if there is one).

## Basic Features ##

* Per-user settings.
* Multisite support.

## Installation ##

This section describes how to install the plugin and get it working.

1. Upload the plugin files to the `/wp-content/plugins/sticky-postbox` directory, or install the plugin through the WordPress _Plugins_ screen directly.
1. Activate the plugin through the _Plugins_ screen in WordPress.

## Usage ##

Once the plugin is installed and activated you can sticky administration meta boxes using sticky icon (it appears on the top right corner of each box).

## API ##

### Constants ###

You can use define following constants in your _wp-config.php_ file.

**`STICKY_POSTBOX_DEBUG`**

Turn on debug messages (also `WP_DEBUG` has to be `true`).
Default to `false`.

**`STICKY_POSTBOX_GLOBAL_OPTIONS`**

In a Multisite installation, allows to shares user's sticky meta boxes settings across all his sites. For example, if you sticky *Publish* box in editing Pages in a specific site, it will be sticky on all Pages of all sites.
Default to `false`.

**`STICKY_POSTBOX_AUTOENABLE`**

By default if the plugin has been activated, it starts automatically. 
Define to `false` if you want to init it manually, for example:

	// Activate manually the plugin once all plugin have been loaded.
	add_action( 'plugins_loaded', function() {
		\Sticky_Postbox::get_instance( 
			array(
				'debug'          => WP_DEBUG,
				'global_options' => false,
			)
		);
	} );

### Hooks ###

Currently the plugin doesn't have actions or filters.

## Frequently Asked Questions ##

### Does it work with Gutenberg? ###

This plugin is intended to work with Classic Editor and Dashboards meta boxes.

### Does it work with multisite installation? ###

Yes

## Screenshots ##

### 1. Sticky meta box in classic edit post. ###
![Sticky meta box in classic edit post.](https://raw.githubusercontent.com/enrico-sorcinelli/sticky-postbox/master/assets-wp/screenshot-1.png)

### 2. Sticky meta box in Dashboard. ###
![Sticky meta box in Dashboard.](https://raw.githubusercontent.com/enrico-sorcinelli/sticky-postbox/master/assets-wp/screenshot-2.png)


## Changelog ##

For plugin changelog, please see [the Releases page on GitHub](https://github.com/enrico-sorcinelli/sticky-postbox/releases).



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.3.0] - 2022-07-12
### Changed
 - Documentation improvements.

### Fixed
 - Input sanitization.

## [1.2.0] - 2022-07-10
### Added 
 - `STICKY_POSTBOX_GLOBAL_OPTIONS` constant.
 - Handled meta boxes on network Dashboard.

### Changed
 - Documentation improvements.

## [1.1.0] - 2022-07-09
### Added 
 - Screenshots and icons.

### Changed
 - Documentation improvements.

### Fixed
 - Screen meta

## [1.0.0] - 2022-07-08
### Added
 - Initial release


tests/phpunit/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package sticky-postbox
 */

$_tests_dir = getenv( 'WP_TESTS_DIR' );

if ( ! $_tests_dir ) {
	$_tests_dir = rtrim( sys_get_temp_dir(), '/\\' ) . '/wordpress-tests-lib';
}

if ( ! file_exists( $_tests_dir . '/includes/functions.php' ) ) {
	echo "Could not find $_tests_dir/includes/functions.php, have you run bin/install-wp-tests.sh ?" . PHP_EOL;
	exit( 1 );
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '../../sticky-postbox.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


tests/phpunit/tests/test-base.php
<?php
/**
 * Class Test_Base
 *
 * @package sticky-postbox
 */

/**
 * Sample test case.
 */
class Test_Base extends WP_UnitTestCase {

	/**
	 * A single example test.
	 */
	function test_sample() {
		// Replace this with some actual testing code.
		$this->assertTrue( true );
	}
}


Gruntfile.js
/* jshint node:true */
module.exports = function( grunt ) {
	'use strict';

	var options = {
		phpcs_bin: grunt.option( 'phpcs-bin' ) || 'phpcs'
	};

	grunt.initConfig({

		// Load package data.
		pkg: grunt.file.readJSON( 'package.json' ),

		// Set folder vars.
		dirs: {
			css: 'assets/css',
			js: 'assets/js',
			languages: 'languages',
			build: 'build'
		},

		// Javascript linting with jshint.
		jshint: {
			options: {
				jshintrc: '.jshintrc',
                reporterOutput: ''
			},
			all: [
				'Gruntfile.js',
				'<%= dirs.js %>/*.js',
				'!**/*.min.js'
			]
		},

		// Javascript linting with eslint.
		eslint: {
			options: {
				configFile: '.eslintrc.json'
			},
			target: [
				'Gruntfile.js',
				'<%= dirs.js %>/**/*.js',
				'!**/*.min.js'
			]
		},

		// Minify .js files.
		uglify: {
			options: {

				//preserveComments: 'some',
				banner: '/* <%= pkg.name %> */\n'
			},
			admin: {
				files: [ {
					expand: true,
					cwd: '<%= dirs.js %>/',
					src: [
						'*.js',
						'!*.min.js'
					],
					dest: '<%= dirs.js %>/',
					ext: '.min.js'
				} ]
			}
		},

		// CSS linting with stylelint.
		stylelint: {
			options: {
				configFile: '.stylelintrc.json'
			},
			all: [
				'<%= dirs.css %>/**/*.css',
				'!**/*.min.css'
			]
		},

		// Minify .css files.
		cssmin: {
			minify: {
				options: {
					banner: '/* <%= pkg.title %> */'
				},
				files: [ {
					expand: true,
					cwd: '<%= dirs.css %>/',
					src: [
						'*.css',
						'!*.min.css'
					],
					dest: '<%= dirs.css %>/',
					ext: '.min.css'
				} ]
			}
		},

		// PHP Code Sniffer.
		phpcs: {
			application: {
				src: [
					'**/*.php',
					'!node_modules/**',
					'!<%= dirs.build %>/**'
				]
			},
			options: {
				bin: options.phpcs_bin + ' --exclude=Generic.Files.LineEndings',
				standard: 'WordPress-Extra',
				verbose: true
			}
		},

		// Watch changes for assets.
		watch: {
			js: {
				files: [ '<%= dirs.js %>/*.js' ],
				tasks: [ 'uglify' ],
				options: {
					spawn: false
				}
			},
			css: {
				files: [ '<%= dirs.css %>/*.css' ],
				tasks: [ 'cssmin' ],
				options: {
					spawn: false
				}
			},
			readme: {
				files: [ 'readme.txt' ],
				tasks: [ 'wp_readme_to_markdown' ],
				options: {
					spawn: false
				}
			}
		},

		// Clean build dir.
		clean: {
			main: [ '<%= dirs.build %>/<%= pkg.name %>' ]
		},

		// Copy the plugin to a versioned release directory.
		copy: {
			main: {
				src: [
					'**',
					'!.git/**',
					'!.gitignore',
					'!.gitmodules',
					'!.jshintrc',
					'!.scrutinizer.yml',
					'!node_modules/**',
					'!<%= dirs.build %>/**',
					'!Gruntfile.js',
					'!package.json',
					'!package-lock.json',
					'!composer.json',
					'!LICENSE',
					'!README.md',
					'!CONTRIBUTING.md',
					'!CHANGELOG.md',
					'!TODO.md',
					'!nbproject/**',
					'!**/*.LCK',
					'!**/_notes/**',
					'!tmp/**',
					'!assets-wp/**',
					'!bin/**',
					'!tests/**',
					'!Makefile',
					'!phpcs.xml',
					'!phpunit.xml   ',
					'!wp-cli/wp-cli-test*'
				],
				dest: '<%= dirs.build %>/<%= pkg.name %>/'
			}
		},

		// Convert line endings to LF.
		lineending: {
			build: {
				options: {
					eol: 'lf',
					overwrite: true
				},
				files: [ {
					expand: true,
					cwd: '<%= dirs.build %>/<%= pkg.name %>/',
					src: [ '**/*.{php,css,js,po,txt}' ]
				} ]
			}
		},

		// Create zip package.
		compress: {
			main: {
				options: {
					mode: 'zip',
					archive: './<%= dirs.build %>/<%= pkg.name %>-<%= pkg.version %>.zip'
				},
				expand: true,
				cwd: '<%= dirs.build %>/<%= pkg.name %>/',
				src: [ '**/*' ],
				dest: '<%= pkg.name %>/'
			}
		},

		// Generate .pot file.
		makepot: {
			target: {
				options: {
					domainPath: '/languages',
					exclude: [
						'<%= dirs.build %>/.*',
						'tmp/.*'
					],
					potFilename: 'sticky-postbox.pot',
					processPot: function( pot ) {
						var translation, deleteTranslation,
							excludedStrings = [ 'Title:', 'Visual', 'HTML', 'Cheatin&#8217; uh?', 'Automatically add paragraphs' ],
							excludedMeta = [ 'Plugin Name of the plugin/theme', 'Plugin URI of the plugin/theme', 'Author of the plugin/theme', 'Author URI of the plugin/theme' ];
						pot.headers['report-msgid-bugs-to'] = 'https://github.com/enrico-sorcinelli/sticky-postbox/issues\n';
						pot.headers['plural-forms'] = 'nplurals=2; plural=n != 1;';
						pot.headers['last-translator'] = 'Enrico Sorcinelli\n';
						pot.headers['language-team'] = 'Enrico Sorcinelli\n';
						pot.headers['x-poedit-basepath'] = '.\n';
						pot.headers['x-poedit-language'] = 'English\n';
						pot.headers['x-poedit-country'] = 'United States\n';
						pot.headers['x-poedit-sourcecharset'] = 'utf-8\n';
						pot.headers['x-poedit-keywordslist'] = '__;_e;__ngettext:1,2;_n:1,2;__ngettext_noop:1,2;_n_noop:1,2;_c,_nc:4c,1,2;_x:1,2c;_ex:1,2c;_nx:4c,1,2;_nx_noop:4c,1,2;esc_html__;esc_html_e;esc_html_x;esc_attr__;esc_attr_e;esc_attr_x;\n';
						pot.headers['x-poedit-bookmarks'] = '\n';
						pot.headers['x-poedit-searchpath-0'] = '.\n';
						pot.headers['x-textdomain-support'] = 'yes\n';
						for ( translation in pot.translations['']) {
							deleteTranslation = false;
							if ( 0 <= excludedStrings.indexOf( translation ) ) {
								deleteTranslation = true;
								console.log( 'Excluded string: ' + translation );
							}
							if ( 'undefined' !== typeof pot.translations[''][translation].comments.extracted ) {
								if ( 0 <= excludedMeta.indexOf( pot.translations[''][translation].comments.extracted ) ) {
									deleteTranslation = true;
									console.log( 'Excluded meta: ' + pot.translations[''][translation].comments.extracted );
								}
							}
							if ( deleteTranslation ) {
								delete pot.translations[''][translation];
							}
						}
						return pot;
					},
					type: 'wp-plugin',
					updateTimestamp: true
				}
			}
		},

		// Check plugin text domain.
		checktextdomain: {
			options: {
				text_domain: 'sticky-postbox', // eslint-disable-line camelcase
				keywords: [
					'__:1,2d',
					'_e:1,2d',
					'_x:1,2c,3d',
					'esc_html__:1,2d',
					'esc_html_e:1,2d',
					'esc_html_x:1,2c,3d',
					'esc_attr__:1,2d',
					'esc_attr_e:1,2d',
					'esc_attr_x:1,2c,3d',
					'_ex:1,2c,3d',
					'_n:1,2,4d',
					'_nx:1,2,4c,5d',
					'_n_noop:1,2,3d',
					'_nx_noop:1,2,3c,4d'
				],
				report_missing: false // eslint-disable-line camelcase
			},
			files: {
				src: [
					'sticky-postbox.php',
					'**/*.php',
					'!node_modules/**',
					'!tmp/**',
					'!<%= dirs.build %>/**'
				],
				expand: true
			}
		},

		// Generate .mo files from .po files.
		potomo: {
			dist: {
				options: {
					poDel: false
				},
				files: [ {
					expand: true,
					cwd: '<%= dirs.languages %>',
					src: [ '*.po' ],
					dest: '<%= dirs.languages %>',
					ext: '.mo',
					nonull: true
				} ]
			}
		},

		// Generate README.md from readme.txt.
		wp_readme_to_markdown: { // eslint-disable-line camelcase
			readme: {
				files: {
					'README.md': 'readme.txt'
				},
				options: {
					screenshot_url: 'https://raw.githubusercontent.com/enrico-sorcinelli/{plugin}/master/assets-wp/{screenshot}.png' // eslint-disable-line camelcase
				}
			}
		},

		// Check version.
		checkwpversion: {
			options: {
				readme: 'readme.txt',
				plugin: 'sticky-postbox.php'
			},
			pluginVsReadme: { //Check plugin header version against stable tag in readme.
				version1: 'plugin',
				version2: 'readme',
				compare: '=='
			},
			pluginVsGrunt: { //Check plugin header version against package.json version.
				version1: 'plugin',
				version2: '<%= pkg.version %>',
				compare: '=='
			},
			pluginVsInternal: { //Check plugin header version against internal defined version.
				version1: 'plugin',
				version2: grunt.file.read( 'sticky-postbox.php' ).match( /VERSION', '(.*)'/ )[1],
				compare: '=='
			},
			pluginVsCHANGES: {
				version1: 'plugin',
				version2: grunt.file.read( 'CHANGELOG.md' ).match( /## \[(.*)\]/ )[1],
				compare: '=='
			}
		}

	});

	// Load NPM tasks to be used here.
	require( 'load-grunt-tasks' )( grunt );

	// Register tasks.
	grunt.registerTask( 'default', [
		'cssmin',
		'uglify'
	]);

	grunt.registerTask( 'check', [
		'jshint',
		'eslint',
		'stylelint',
		'checktextdomain',
		'checkwpversion'
	]);

	grunt.registerTask( 'languages', [
		'checktextdomain',
		'makepot',
		'potomo'
	]);

	grunt.registerTask( 'readme', [
		'wp_readme_to_markdown'
	]);

	grunt.registerTask( 'build', [
		'check',
		'default',
		'readme',
		'clean',
		'copy',
		'lineending',
		'compress'
	]);

};


assets/js/admin.js
/**
 * @file This file contains Sticky Postbox JavaScript
 * @author Enrico Sorcinelli
 * @version 1.1.0
 * @title Sticky Postbox
 */

// Make all in a closure.
;( function( $ ) {

	$( document ).ready( function() {

		/**
		 * Handle sticky postboxes.
		 *
		 */
		window.stickyPostbox = function( args ) {

			// Set defaults.
			args = $.extend( true, {
				postbox: ''
			}, args );

			var toggle_sticky = function( element ) {

				var $postbox = $( element ).closest( '.postbox' );

				$postbox.toggleClass( 'sticky-postbox-sticky' );

				// Turn on/off sortable.
				$( element ).siblings( 'h2' ).toggleClass( 'hndle' );
				$( element ).parent().siblings( 'h2' ).toggleClass( 'hndle' );

				// Add compatibility CSS.
				if ( /^dashboard(-network)?$/.test( pagenow ) ) {
					$( element ).siblings( 'h2' ).toggleClass( 'sticky-postbox-hndle-compat' );
					$( element ).parent().siblings( 'h2' ).toggleClass( 'sticky-postbox-hndle-compat' );
				}
				$( element ).siblings( 'h2' ).toggleClass( 'sticky-postbox-hndle-compat-border' );

				// Toogle order buttons visibility.
				$( element ).siblings( 'button.handle-order-higher, button.handle-order-lower' ).toggle();

				// Put screen-meta info over sticky postbox.
				if ( $postbox.hasClass( 'sticky-postbox-sticky' ) ) {
					$( '#screen-meta, .screen-meta-toggle' ).addClass( 'sticky-postbox-screen-meta' );
				} else {
					$( '#screen-meta, .screen-meta-toggle' ).removeClass( 'sticky-postbox-screen-meta' );
				}
			};

			// Add sticky icon.
			$( args.postbox + '.postbox button.handlediv' ).after( '<span class="dashicons sticky-postbox-icon-sticky"></span>' );

			// Add click handler.
			$( args.postbox + '.postbox .sticky-postbox-icon-sticky' ).on( 'click', function() {

				var sticky_postboxes = [];

				// Only one sticky box at once.
				if ( ! $( this ).closest( '.postbox' ).hasClass( 'sticky-postbox-sticky' ) ) {
					toggle_sticky( $( '.postbox.sticky-postbox-sticky .sticky-postbox-icon-sticky' ) );
				}

				toggle_sticky( this );

				// Get sticky postbox.
				$( '.postbox.sticky-postbox-sticky .sticky-postbox-icon-sticky' ).closest( '.postbox' ).each( function() {
					sticky_postboxes.push( $( this ).attr( 'id' ) );
				});

				// Save status doing AJAX query.
				$.ajax({
					url: ajaxurl,
					type: 'post',
					timeout: 1000 * args.timeout,
					dataType: 'json',
					data: {
						action: 'sticky_postbox_sticky_postboxes',
						page: typenow || pagenow,
						_ajax_nonce_sticky_postbox_sticky_postboxes: sticky_postbox_i18n.nonces.sticky_postboxes,
						sticky: sticky_postboxes,
						is_network_admin: sticky_postbox_i18n.is_network_admin
					}
				});

			});

			// Restore sticky postboxes.
			if ( 'undefined' !== typeof sticky_postbox_i18n.sticky_postboxes ) {
				sticky_postbox_i18n.sticky_postboxes.forEach(
					function( el ) {
						$( '#' + el  +  ' .sticky-postbox-icon-sticky' ).trigger( 'click' );
					}
				);
			}

		};

		stickyPostbox();

	});

})( jQuery );


php/class-sticky-postbox.php
<?php
/**
 * Plugin base class.
 *
 * @package sticky-postbox
 * @author Enrico Sorcinelli
 */

// Check running WordPress instance.
if ( ! defined( 'ABSPATH' ) ) {
	header( 'HTTP/1.1 404 Not Found' );
	exit();
}

/**
 * Plugin base class.
 */
class Sticky_Postbox {

	/**
	 * Instance of this class.
	 *
	 * @var object
	 */
	public static $instance;

	/**
	 * For use with minified libraries.
	 *
	 * @var string
	 */
	private $assets_suffix;

	/**
	 * Instance settings.
	 *
	 * @var array
	 */
	private $settings = false;

	/**
	 * Class constructor.
	 *
	 * @param array $args {
	 *     Argument list.
	 *     @type boolean $debug Debug mode.
	 * }
	 */
	public function __construct( $args = array() ) {

		$this->settings = wp_parse_args(
			$args,
			array(
				'debug'          => false,
				'global_options' => false,
			)
		);

		// Load plugin text domain.
		load_plugin_textdomain( 'sticky-postbox', false, dirname( plugin_basename( __FILE__ ) ) . '/../languages/' );

		// Admin init.
		add_action( 'admin_init', array( $this, 'admin_init' ) );
	}

	/**
	 * Get the singleton instance of this class.
	 *
	 * @param array $args Constructor arguments list.
	 *
	 * @return object
	 */
	public static function get_instance( $args = array() ) {
		if ( ! ( self::$instance instanceof self ) ) {
			self::$instance = new self( $args );
		}
		return self::$instance;
	}

	/**
	 * Admin init hook.
	 *
	 * @return void
	 */
	public function admin_init() {

		// Use minified libraries if SCRIPT_DEBUG is turned off.
		$this->assets_suffix = ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ) ? '' : '.min';

		// AJAX action for user sticky settings.
		add_action( 'wp_ajax_sticky_postbox_sticky_postboxes', array( $this, 'ajax_sticky_postboxes' ) );

		// Add CSS and JS to right pages.
		foreach ( array( 'index.php', 'post.php', 'post-new.php' ) as $page ) {
			add_action( 'admin_print_styles-' . $page, array( $this, 'load_css' ), 10, 0 );
			add_action( 'admin_print_scripts-' . $page, array( $this, 'load_javascript' ), 10, 0 );
		}
	}

	/**
	 * Load CSS files.
	 *
	 * @return void
	 */
	public function load_css() {
		wp_enqueue_style(
			'sticky-postbox-css',
			STICKY_POSTBOX_BASEURL . '/assets/css/admin' . $this->assets_suffix . '.css',
			array(),
			STICKY_POSTBOX_VERSION
		);
	}

	/**
	 * Load JavaScript files.
	 *
	 * @return void
	 */
	public function load_javascript() {
		global $pagenow;

		list( $current_post, $post_type ) = self::get_current_post_and_post_type();

		$user_id = get_current_user_id();

		// Sticky postboxes handling.
		if ( preg_match( '/^(index|post(-new)?)\.php$/', $pagenow ) ) {

			$page = preg_match( '/^index.php$/', $pagenow )
				? ( 'dashboard' . ( is_network_admin() ? '-network' : '' ) )
				: $post_type;

			$sticky_postboxes = is_network_admin() && false === $this->settings['global_options'] ?
				get_user_meta( $user_id, 'sticky_postbox_sticky_postboxes_' . $page, true )
				: get_user_option( 'sticky_postbox_sticky_postboxes_' . $page, $user_id );
		}

		if ( empty( $sticky_postboxes ) ) {
			$sticky_postboxes = array();
		}

		wp_enqueue_script(
			'sticky-postbox-admin-js',
			STICKY_POSTBOX_BASEURL . '/assets/js/admin' . $this->assets_suffix . '.js',
			array(),
			STICKY_POSTBOX_VERSION,
			false
		);

		// Localization.
		wp_localize_script(
			'sticky-postbox-admin-js',
			'sticky_postbox_i18n',
			array(
				'nonces'           => array(
					'sticky_postboxes' => wp_create_nonce( 'sticky_postbox_sticky_postboxes' ),
				),
				'sticky_postboxes' => $sticky_postboxes,
				'msgs'             => array(),
				'is_network_admin' => (bool) is_network_admin(),
			)
		);
	}

	/**
	 * AJAX sticky postboxes handle.
	 */
	public function ajax_sticky_postboxes() {

		// Check nonce code.
		check_ajax_referer( 'sticky_postbox_sticky_postboxes', '_ajax_nonce_sticky_postbox_sticky_postboxes' );

		$args = wp_parse_args(
			$_REQUEST,
			array(
				'sticky'           => array(),
				'page'             => '',
				'is_network_admin' => false,
			)
		);

		if ( sanitize_key( $args['page'] ) !== $args['page'] ) {
			wp_die( 0 );
		}

		if ( ! $user = wp_get_current_user() ) {
			wp_die( -1 );
		}

		// Update status.
		if ( is_array( $args['sticky'] ) ) {
			$ret = update_user_option(
				$user->ID,
				'sticky_postbox_sticky_postboxes_' . $args['page'],
				$args['sticky'],
				false === $this->settings['global_options'] && empty( $args['is_network_admin'] ) ? false : true
			);
			$this->log( $user->ID, $args, $ret );
		}
		wp_die( 1 );
	}

	/**
	 * Try to return current post and post_type.
	 *
	 * @return array
	 */
	public static function get_current_post_and_post_type() {
		global $post;

		$current_post = $post;
		$post_type    = 'post';

		// Try to get current post if it isn't yet defined.
		if ( ! ( $current_post instanceof \WP_Post ) && ( isset( $_REQUEST['post'] ) || isset( $_REQUEST['post_ID'] ) ) ) {
			$current_post = get_post( isset( $_REQUEST['post'] ) ? sanitize_text_field( $_REQUEST['post'] ) : sanitize_text_field( $_REQUEST['post_ID'] ) );
		}

		// Set post type.
		if ( ! empty( $current_post ) ) {
			$post_type = $current_post->post_type;
		}
		elseif ( isset( $_REQUEST['post_type'] ) ) {
			$post_type = sanitize_text_field( $_REQUEST['post_type'] );
		}

		return array( $current_post, $post_type );
	}

	/**
	 * Debugging helper.
	 */
	private function log() {
		if ( false === $this->settings['debug'] ) {
			return;
		}
		error_log( print_r( func_get_args(), true ) );
	}

	/**
	 * Plugin uninstall hook.
	 *
	 * @return void
	 */
	public static function plugin_uninstall() {
	}
}



sticky-postbox.php
<?php
/**
 * Sticky Postbox plugin for WordPress.
 *
 * @package sticky-postbox
 *
 * Plugin Name: Sticky Postbox
 * Plugin URI:  https://wordpress.org/plugins/sticky-postbox
 * Description: A WordPress plugin that allow to sticky administration postboxes.
 * Author:      Enrico Sorcinelli
 * Author URI:  https://github.com/enrico-sorcinelli/sticky-postbox/graphs/contributors
 * Text Domain: sticky-postbox
 * Domain Path: /languages/
 * Version:     1.3.0
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 */

// Check running WordPress instance.
if ( ! defined( 'ABSPATH' ) ) {
	header( 'HTTP/1.1 404 Not Found' );
	exit();
}

// Plugins constants.
define( 'STICKY_POSTBOX_VERSION', '1.3.0' );
define( 'STICKY_POSTBOX_BASEDIR', dirname( __FILE__ ) );
define( 'STICKY_POSTBOX_BASEURL', plugin_dir_url( __FILE__ ) );

// Default values.
foreach( array( 'STICKY_POSTBOX_DEBUG', 'STICKY_POSTBOX_GLOBAL_OPTIONS' ) as $key ) {
	if ( ! defined( $key ) ) {
		define( $key, false );
	}
}

if ( ! class_exists( 'Sticky_Postbox' ) ) {

	require_once STICKY_POSTBOX_BASEDIR . '/php/class-sticky-postbox.php';

	/**
	 * Init the plugin.
	 *
	 * Define STICKY_POSTBOX_AUTOENABLE to `false` in your wp-config.php to disable.
	 */
	function sticky_postbox_init() {

		if ( defined( 'STICKY_POSTBOX_AUTOENABLE' ) && false === STICKY_POSTBOX_AUTOENABLE ) {
			return;
		}

		// Instantiate our plugin class and add it to the set of globals.
		// Create plugin instance object only under administration interface.
		if ( is_admin() || is_network_admin() ) {
			$GLOBALS['sticky_postbox'] = \Sticky_Postbox::get_instance( 
				array( 
					'debug'          => STICKY_POSTBOX_DEBUG && WP_DEBUG,
					'global_options' => STICKY_POSTBOX_GLOBAL_OPTIONS,
				)
			);
		}
	}

	// Activate the plugin once all plugin have been loaded.
	add_action( 'plugins_loaded', 'sticky_postbox_init' );

	// Activation/Deactivation hooks.
	register_uninstall_hook( __FILE__, array( 'Sticky_Postbox', 'plugin_uninstall' ) );
}
